Intel GPU oneDNN upstreaming for primitive integration

ZhiweiYan-96 opened the pull request:
# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

The code in this PR would be used in following PRs, where aten operators would call the functions in these integration code.. We separate the prs due to onednn integration is logically separable with aten operator implementation. Also, this can ease the burden of reviewing by avoid too much codes in single PR.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117112



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

ZhiweiYan-96 committed:
Intel GPU oneDNN upstreaming for primitive integration

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/detail/Attr.h:
```diff
@@ -0,0 +1,370 @@
+    return *this;
+  }
+
+  void extract_post_ops(dnnl::post_ops& dnnl_post_ops, const at::Tensor& dst) {
```
construct a dnnl::post_ops as member of Attr class will be more reasonable and reduce the code line?

ZhiweiYan-96 replied:
Thanks for suggestion, I have changed the API. Attr would hold a member with type `dnnl::post_ops`, and Attr class can construct and return it to users.

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/detail/Attr.h:
```diff
@@ -0,0 +1,370 @@
+  kind_t kind_ = kind_t::eltwise;
+};
+
+class Attr {
```
derive from dnnl::attr should be more reasonable and reduce the code lines?

ZhiweiYan-96 replied:
Yes, derive from `dnnl::attr` would help us align the semantic and follow the api changes. Due to `Attr` change would affect all integration code(`GEMM&Conv` and their fusions), I would need verify the code change with plenty of UTs . So, I may need upload another PR for it with the functionality (Conv&GEMM) steady(can have ut for testing) later.

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/detail/Runtime.h:
```diff
@@ -0,0 +1,103 @@
+#pragma once
```
Is it more reasonable to rename to Context.h or oneDNNContext.h?

ZhiweiYan-96 replied:
Thanks for the suggestion, has renamed the file

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/detail/Runtime.h:
```diff
@@ -0,0 +1,103 @@
+// Keep non-static and non-inline
+bool set_onednn_verbose(int level);
+
+static inline dnnl::memory xpu_onednn_memory(
```
make_onednn_memory?

ZhiweiYan-96 replied:
Have modified.

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/detail/Utils.h:
```diff
@@ -0,0 +1,376 @@
+#define XPU_ONEDNN_EXEC(prim, stream, ...)                           \
+  {                                                                    \
+    auto q = dnnl::sycl_interop::get_queue((stream));                  \
+    XPU_EXT_SUBMIT(                                                  \
```
Will we still keep these macros?

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/detail/Utils.h:
```diff
@@ -0,0 +1,376 @@
+  return true;
+}
+
+static bool is_wrapped_number(const at::Tensor& t) {
```
remove this util, call the pytorch API directly, should be more clear for maintainer?

ZhiweiYan-96 replied:
removed

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/detail/Utils.h:
```diff
@@ -0,0 +1,376 @@
+  return false;
+}
+
+enum MEMORY_LAYOUT_FOR_CONV {
```
Is it necessary to define in integration code?

ZhiweiYan-96 replied:
removed

guangyey commented on aten/src/ATen/native/mkldnn/xpu/detail/Matmul.h:
```diff
@@ -0,0 +1,245 @@
+    result.copy_(dst);
+}
+
+} // namespace onednn
```
Use nested namespace.

ZhiweiYan-96 replied:
Has modified to `at` `native::xpu` `onednn`, thanks for remiding

xiaolil1 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

xiaolil1 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey commented on aten/src/ATen/native/mkldnn/xpu/detail/Attr.h:
```diff
@@ -0,0 +1,370 @@
+            dnnl::memory::format_tag::abcde);
+        break;
+      default:
+        AT_ERROR(
```
`AT_ERROR` is deprated. ` TORCH_INTERNAL_ASSERT` is better.

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

EikanWang commented on aten/src/ATen/native/mkldnn/xpu/detail/Attr.h:
```diff
@@ -0,0 +1,375 @@
+#include <oneapi/dnnl/dnnl_types.h>
+#include <ATen/native/mkldnn/xpu/detail/Utils.h>
+
+namespace at {
+namespace native::xpu {
+namespace onednn{
```
why not `namespace at::native::xpu::onednn`

ZhiweiYan-96 replied:
Has changed to `at::native::onednn`, thanks for suggestions.

EikanWang commented on aten/src/ATen/native/mkldnn/xpu/detail/oneDNNContext.cpp:
```diff
@@ -0,0 +1,26 @@
+#include <ATen/native/mkldnn/xpu/detail/oneDNNContext.h>
+#include <ATen/native/mkldnn/xpu/detail/Utils.h>
+
+/* XXX WARNING:
```
What's the meaning of `XXX WARNING`?

ZhiweiYan-96 replied:
removed

EikanWang commented on aten/src/ATen/native/mkldnn/xpu/detail/Utils.h:
```diff
@@ -0,0 +1,362 @@
+    const at::Tensor& self,
+    const at::Tensor& other,
+    bool is_fusion = false) {
+  // FIXME: update onednn
```
Do we still need to fix it?

ZhiweiYan-96 replied:
no, removed

EikanWang commented on aten/src/ATen/native/mkldnn/xpu/detail/Utils.h:
```diff
@@ -0,0 +1,362 @@
+  }
+
+  auto block_size = 1;
+  // const auto& blk = md->format_desc.blocking;
```
```suggestion
```

ZhiweiYan-96 replied:
removed

EikanWang commented on aten/src/ATen/native/mkldnn/xpu/detail/Matmul.h:
```diff
@@ -0,0 +1,244 @@
+  dnnl::memory m1_m = m1_usr_m, m2_m = m2_usr_m, dst_m = dst_usr_m;
+  at::Tensor m1_, m2_, dst_;
+
+  // TODO: hanld attr.with_sum()
```
```suggestion
  // TODO: handle attr.with_sum()
```

ZhiweiYan-96 replied:
non need anymore, removed

guangyey commented on aten/src/ATen/native/mkldnn/xpu/detail/oneDNNContext.cpp:
```diff
@@ -0,0 +1,26 @@
+ * */
+namespace at::native{
+namespace xpu {
+namespace onednn {
```
namespace `at::native::xpu::onednn` is so long. If the user uses `using namespace at::native;` in his code,  it is easy to have a namespace conflict with `c10::xpu` and `at::xpu` when call `xpu::onednn::func`. Because the compiler doesn't know which `xpu` should be referred to(`c10::xpu`, `at::xpu` or `at::native::xpu`)? Could you simplify this namespace to `at::native::onednn`. The user can call these funcs via `onednn::func` instead of `func` under `using namespace at::native::xpu::onednn;`

ZhiweiYan-96 replied:
Thanks for suggestion, the namespace has been changed to `at::native::onednn`

guangyey commented on aten/src/ATen/native/mkldnn/xpu/detail/oneDNNContext.cpp:
```diff
@@ -0,0 +1,26 @@
+
+} // namespace onednn
+} // namespace xpu
+} // namespace at::native
```
need a blank line.

ZhiweiYan-96 replied:
added

leizhenyuan committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

**Mention is will be used in later PR**




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

The code in this PR would be used in following PRs, where aten operators would call the functions in these integration code.. We separate the prs due to onednn integration is logically separable with aten operator implementation. Also, this can ease the burden of reviewing by avoid too much codes in single PR.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/detail/Utils.h:
```diff
@@ -0,0 +1,360 @@
+
+#define XPU_ONEDNN_EXEC(prim, stream, ...)                               \
+{                                                                        \
+  auto e = dnnl::sycl_interop::execute((prim), (stream), ##__VA_ARGS__); \
```
As long as the e is created in {}, it will be useless for rest of op impl, right?

ZhiweiYan-96 replied:
Yes, I would remove this marco and directly call execute in primitive. I think this marco is not needed anymore.

ZhiweiYan-96 replied:
hi, @gujinghui , I have removed this marco, the primitive would directly use `dnnl::sycl_interop::execute` for calling primitive. 

Also, I have change the matmul and convolution API, which would expose `sycl::event` to user just like the codes in IPEX.  You may refer to 

https://github.com/pytorch/pytorch/pull/117112/files#diff-b3a3b45f243b153b61b63b33bee989ef2ecd7f5b28970555adfc7d4a92b35b81R11, 

and conv in #117512

gujinghui replied:
Looks great. BTW, will you append below APIs into matmul integration layer?
https://github.com/intel-innersource/frameworks.ai.pytorch.ipex-gpu/pull/3926

ZhiweiYan-96 replied:
Yes, I may put it in another separate PR, current PRs would focus on the high priority ops, like addbmm, bmm.

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/detail/oneDNNContext.h:
```diff
@@ -0,0 +1,81 @@
+// Keep non-static and non-inline
+bool set_onednn_verbose(int level);
+
+static inline dnnl::memory make_onednn_memory(
```
we need these APIs to be public, if considering to reuse them in IPEX, right?

ZhiweiYan-96 replied:
We need to add TORCH_API for it?

ZhiweiYan-96 replied:
Sorry that I may need further explanation on this. 
1. If the problem is the `static inline` declaration, I think the user could directly `#include <detail/oneDNN>` or `#include <detail/oneDNNContext.h>` like our op in aten level?
2. If the concern lies at linkage, so we need to put it into a `cpp` source file and declared as `TOCH_API`?

gujinghui replied:
Why we do not move all impl in this file to corresponding ccp file, then, expose the APIs by TORCH_API?

ZhiweiYan-96 replied:
A commit has been pushed, 
1. Runtime related API is exposed, like `xpu_onednn_memory`, `GpuEngineManger`, `GpuStreamManager`
2. Function in `Utils.h` is mostly a part of onednn integration specific code, users could easily rewrite in their repos. I suggest we do not export as an API of TORCH.

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

The code in this PR would be used in following PRs, where aten operators would call the functions in these integration code.. We separate the prs due to onednn integration is logically separable with aten operator implementation. Also, this can ease the burden of reviewing by avoid too much codes in single PR.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

The code in this PR would be used in following PRs, where aten operators would call the functions in these integration code.. We separate the prs due to onednn integration is logically separable with aten operator implementation. Also, this can ease the burden of reviewing by avoid too much codes in single PR.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/detail/oneDNNContext.h:
```diff
@@ -0,0 +1,81 @@
+  ~GpuStreamManager() {}
+
+};
+
```
Can we add and expose below util API, as well?
https://github.com/intel-innersource/frameworks.ai.pytorch.ipex-gpu/blob/26be453c0b31fea28f35a282a5e1878bd3b74440/csrc/gpu/oneDNN/Runtime.cpp#L24

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

The code in this PR would be used in following PRs, where aten operators would call the functions in these integration code.. We separate the prs due to onednn integration is logically separable with aten operator implementation. Also, this can ease the burden of reviewing by avoid too much codes in single PR.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

The code in this PR would be used in following PRs, where aten operators would call the functions in these integration code.. We separate the prs due to onednn integration is logically separable with aten operator implementation. Also, this can ease the burden of reviewing by avoid too much codes in single PR.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for primitive integration"

# Motivation

As proposed in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem.


Current PR is based on #117098, where oneDNN library for Intel GPU should be ready.  This PR is the integration code from aten to oneDNN. GEMM integration code is the core part in this PR. Accompanied with GEMM, more basic support like runtime (device, stream), primitive attr is also included.

We put the oneDNN integration code in directory `aten/src/ATen/native/mkldnn/xpu/detail`. We add a namespace `at::native::xpu::onednn` for oneDNN integration.

The code in this PR would be used in following PRs, where aten operators would call the functions in these integration code.. We separate the prs due to onednn integration is logically separable with aten operator implementation. Also, this can ease the burden of reviewing by avoid too much codes in single PR.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

albanD reviewed (approved):
I guess you build is glob based and so doesn't need updating?
Also not sure what is the test state?

But stamping since this is xpu-only code

ZhiweiYan-96 commented:
Hi, @albanD Thanks for your comments!
I have put all compilation changes in #117202, the integration code compilation is triggered through caller from aten ops like `addmm, addbmm`, etc.  Variable required like `ATen_XPU_SRCS`, `Caffe2_XPU_SRCS` is set in it. The correctness is validated through unit test in #117202 also.

We separate the PR mainly for easier reviewing considering they have big changes but can be logically separated, where this PR is for oneDNN integration, #117202 are for aten operaors impl accompanied testing and compilation.

EikanWang commented:
@albanD , sincerely thanks for your comments. As @ZhiweiYan-96 mentioned, we have ported some operations through copy-paste and passed all the related cases. In parallel, I submitted another PR https://github.com/pytorch/pytorch/pull/122953 to refine the test cases reusing. W/ #122953, we simplify the test reusing work. It would be appreciated if you could review #122953

ZhiweiYan-96 commented:
@pytorchbot  merge

ZhiweiYan-96 commented:
@pytorchbot rebase -b main

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 commented:
@pytorchbot rebase -b main

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -i

ZhiweiYan-96 commented:
@pytorchbot merge
